### PR TITLE
New unreachable token value warning

### PR DESCRIPTION
**Pull request by KvanTTT**
Partial fix for #1072 (see unit tests).

I tested my version on [grammars-v4]() repository and found a couple of such warnings:

* warning(184): [LessLexer.g4:224:0](https://github.com/antlr/grammars-v4/blob/master/less/LessLexer.g4#L224): One of the token FORMAT values unreachable. % is always overlapped by token PERC
* warning(184): [ZLexer.g4:478:0](https://github.com/antlr/grammars-v4/blob/master/z/ZLexer.g4#L478): One of the token ID0 values unreachable. _ is always overlapped by token ARGUMENT

Also I discovered and fixed such warnings in ANTLR unit tests itself (testInvalidLexerCommand, testLexerCommandArgumentValidation).

### CODE REVIEW

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 1]
```diff
@@ -96,8 +89,7 @@ public void checkActionRedefinitions(List<GrammarAST> actions) {
 			nameNode = (GrammarAST) ampersandAST.getChild(0);
 			if (ampersandAST.getChildCount() == 2) {
 				name = nameNode.getText();
-			}
-			else {
+			} else {
```
Comment (parrt) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: hi. please use same style. `}` on line by themselves

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 2]
```diff
@@ -108,29 +100,21 @@ public void checkActionRedefinitions(List<GrammarAST> actions) {
...
-	}
-
-    /** Make sure a label doesn't conflict with another symbol.
-     *  Labels must not conflict with: rules, tokens, scope names,
-     *  return values, parameters, and rule-scope dynamic attributes
-     *  defined in surrounding rule.  Also they must have same type
-     *  for repeated defs.
-     */
-    public void checkForLabelConflicts(Collection<Rule> rules) {
+	/**
```
Comment (parrt) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: are you changing whitespace by mistake?

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 3]
```diff
@@ -341,42 +318,130 @@ public void checkForModeConflicts(Grammar g) {
 		}
 	}
 
+	public void checkForUnreachableTokens(Grammar g) {
```
Comment (parrt) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: can you explain in a comment your strategy for finding unreachable tokens?  I also need to review that strategy to know if it is correct. :)

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java (hunk 2 above)
Comment (KvanTTT) [reply tool/src/org/antlr/v4/semantics/SymbolChecks.java]: Сonversely I replaced all whitespaces on tabs (respect to general project style).

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java (hunk 3 above)
Comment (KvanTTT) [reply tool/src/org/antlr/v4/semantics/SymbolChecks.java]: Done :)

**Comment by KvanTTT** [comment]
Wait, do not merge. I want to implement yet another check.

**Comment by KvanTTT** [comment]
Probably the complexity can be reduced down to linear. I'll take a look until the end of the week.

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 4]
```diff
@@ -223,11 +207,11 @@ private void checkForTypeMismatch(Rule r, LabelElementPair prevLabelPair, LabelE
 					labelPair.type + "!=" + prevLabelPair.type);
 		}
 		if (!prevLabelPair.element.getText().equals(labelPair.element.getText()) &&
-			(prevLabelPair.type.equals(LabelType.RULE_LABEL) || prevLabelPair.type.equals(LabelType.RULE_LIST_LABEL)) &&
-			(labelPair.type.equals(LabelType.RULE_LABEL) || labelPair.type.equals(LabelType.RULE_LIST_LABEL))) {
+				(prevLabelPair.type.equals(LabelType.RULE_LABEL) || prevLabelPair.type.equals(LabelType.RULE_LIST_LABEL)) &&
```
Comment (sharwell) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: :memo: Previous indentation was correct here:

https://github.com/antlr/antlr4/blob/64ae36b3e8a8fe326c38d8739ed44ff728f626d5/runtime/Java/nb-configuration.xml#L36

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 5]
```diff
@@ -308,25 +292,25 @@ protected void checkLocalConflictingDeclarations(Rule r, AttributeDict attribute
...
+					key,
+					r.name);
 		}
 	}
 
 	protected void checkReservedNames(Collection<Rule> rules) {
 		for (Rule rule : rules) {
 			if (reservedNames.contains(rule.name)) {
-				errMgr.grammarError(ErrorType.RESERVED_RULE_NAME, g.fileName, ((GrammarAST)rule.ast.getChild(0)).getToken(), rule.name);
+				errMgr.grammarError(ErrorType.RESERVED_RULE_NAME, g.fileName, ((GrammarAST) rule.ast.getChild(0)).getToken(), rule.name);
```
Comment (sharwell) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: :memo: Previous spacing was correct here:

https://github.com/antlr/antlr4/blob/64ae36b3e8a8fe326c38d8739ed44ff728f626d5/runtime/Java/nb-configuration.xml#L43

File: tool/src/org/antlr/v4/semantics/SymbolChecks.java [hunk 6]
```diff
@@ -341,42 +325,156 @@ public void checkForModeConflicts(Grammar g) {
...
 	public void checkRuleArgs(Grammar g, List<GrammarAST> rulerefs) {
-		if ( rulerefs==null ) return;
+		if (rulerefs == null) return;
 		for (GrammarAST ref : rulerefs) {
 			String ruleName = ref.getText();
 			Rule r = g.getRule(ruleName);
-			GrammarAST arg = (GrammarAST)ref.getFirstChildWithType(ANTLRParser.ARG_ACTION);
-			if ( arg!=null && (r==null || r.args==null) ) {
+			GrammarAST arg = (GrammarAST) ref.getFirstChildWithType(ANTLRParser.ARG_ACTION);
+			if (arg != null && (r == null || r.args == null)) {
```
Comment (sharwell) [inline tool/src/org/antlr/v4/semantics/SymbolChecks.java]: :memo: Spacing here was generally unspecified (@parrt uses the previous spacing, while I use the updated spacing). Generally accepted practice became not changing it unless you were rewriting the line.

File: tool/src/org/antlr/v4/tool/ErrorType.java [hunk 7]
```diff
@@ -522,15 +522,15 @@
 	 */
 	USE_OF_BAD_WORD(134, "symbol <arg> conflicts with generated code in target language or runtime", ErrorSeverity.ERROR),
 	/**
-	 * Compiler Error 134.
+	 * Compiler Error 183.
```
Comment (sharwell) [inline tool/src/org/antlr/v4/tool/ErrorType.java]: 💭 Wasn't this part of another pull request?

File: tool-testsuite/test/org/antlr/v4/test/tool/TestSymbolIssues.java [hunk 8]
```diff
@@ -401,4 +401,34 @@ public void testLabelsForTokensWithMixedTypesLRWithoutLabels() {
...
+				"TOKEN1: 'as' 'df' | 'qwer';\n" +
+				"TOKEN2: [0-9];\n" +
+				"TOKEN3: 'asdf';\n" +
+				"TOKEN4: 'q' 'w' 'e' 'r' | A;\n" +
+				"TOKEN5: 'aaaa';\n" +
+				"TOKEN6: 'asdf';\n" +
+				"TOKEN7: 'qwer'+;\n" +
+				"TOKEN8: 'a' 'b' | 'b' | 'a' 'b';\n" +
+				"fragment\n" +
+				"TOKEN9: 'asdf' | 'qwer' | 'qwer';\n" +
```
Comment (sharwell) [inline tool-testsuite/test/org/antlr/v4/test/tool/TestSymbolIssues.java]: :bulb: Should include tokens with characters that need escapes, to make sure the values are properly escaped in the error messages.

File: tool/src/org/antlr/v4/tool/ErrorType.java (hunk 7 above)
Comment (KvanTTT) [reply tool/src/org/antlr/v4/tool/ErrorType.java]: Yes, it is.

File: tool-testsuite/test/org/antlr/v4/test/tool/TestSymbolIssues.java (hunk 8 above)
Comment (KvanTTT) [reply tool-testsuite/test/org/antlr/v4/test/tool/TestSymbolIssues.java]: Done.

**Comment by parrt** [comment]
Ok, are we ready to go with this merge?

**Comment by KvanTTT** [comment]
@sharwell please check fixes.

**Comment by parrt** [comment]
@KvanTTT How many grammars have you run this on?

**Comment by KvanTTT** [comment]
All grammars from an official [grammars-v4](https://github.com/antlr/grammars-v4) repository, see my first comment: https://github.com/antlr/antlr4/pull/2066#issue-267463865

**Pull request merged**